### ISSUE TITLE
Examples

### ISSUE THREAD

**Issue by toncho11** [rafael2k/8086-toolchain#16]
Awesome job for the complier!!!!!!!!!!!
Thank you @rafael2k and @ghaerr!

1) Please add to the readme how to natively compile an example. It should be with the ecc or the new make or command by command. It could be a wiki page also. Actually the ecc script should be rather part of this repository?
2) Please make a zip file as a github release with the native ELKS tools. An hd ELKS image (fat image is better for Windows users) with the toolchain will also be nice :)

**Comment by ghaerr** [comment]
> Actually the ecc script should be rather part of this repository?

Things are a bit complicated at the moment, as there's no C library at all, only system calls, and most of the .h header files are on ELKS in the C library, but others specific to C86 are in the toolchain libc/include directory. We haven't solidified on a plan for exactly where all the header files and binary components should live just yet. Instead the focus is on seeing what C86 and other tools can and can't do.

That said, yes, ecc should at some point move here, but I've been using it on ELKS (along with ewcc/ewlink) and @rafael2k has been building and testing instead using make within the toolchain examples/ directory.

It's also a bit early to try to make a binary distribution of all this, since it would also have to include the header files from ELKS and toolchain, as well as both the OWC libc.a which isn't automatically built from ELKS, as well as the toolchain libc86.a. Even with a binary distribution, there's no C library so almost no programs can actually be built that will do anything yet  :)

If you want to test a bit, perhaps for now try cloning this repo, build it using "make", then cd examples/ and run make from there, which will get you to see what @rafael2k is playing with.

There will be a binary distribution, but its not quite ready yet! :)

**Comment by rafael2k** [comment]
Hi @toncho11, I'll add instructions to the README. As @ghaerr wrote, we are still somewhat barefoot without libc, but with the system calls many things can be done already. But I'll prepare an image or "supplemental disk" for myself with the toolchain, headers and bare-foot libc to test bare metal, and I'll share it soon.

**Comment by rafael2k** [comment]
Btw, I committed https://github.com/rafael2k/8086-toolchain/issues/16 a simple chess game in the examples, in order there can be exampleS, and something cool to play with natively. Also, I found out nanoprintf was not really working well, with integer conversion broken...

**Comment by ghaerr** [comment]
> I found out nanoprintf was not really working well, with integer conversion broken...

What exact "integer conversion" did you find was broken? The C86 pointer increment bug which caused a register allocation failure, or something else?

I would be interested in learning about exactly what didn't work in nanoprintf.

BTW, I added stddef.h so that no longer needs to be commented out. Still working on // in CPP (can you give specific non-working example?) as well as adding intptr_t and uintptr_t.

**Comment by rafael2k** [comment]
It explodes with a floating point error. I'm trying to pinpoint the line.

**Comment by ghaerr** [comment]
> It explodes with a floating point error. I'm trying to pinpoint the line.

You mean the host compiler quits, or ELKS terminates the process and the shell says something?

I notice that C86 can produce float output, but that's currently turned off. I'll look into this after you find more info. Now that I've gotten this huge memory allocation stuff mostly working, next I plan on looking at fixing the C86 register allocation hang.

**Comment by rafael2k** [comment]
ELKS process terminates with floating point exception when using the %d of nanoprintf.

**Comment by ghaerr** [comment]
I see. Also nanoprintf.h is ridiculously complicated with regards to float conversion, I see it could be a big mess. It is possible that one of your defines is turned off and C86 is outputting some 8086 float instructions, which is causing the problem. Perhaps there is a quick way of looking at the .asm file for float instructions, but I thought you removed those instructions from the slimmed-down NASM, or are they still in there?

**Comment by rafael2k** [comment]
Float instructions are in nasm, but I'm not sure about c86. I see lots of FLOAT_SUPPORT #ifs in c86 source, but looking at check.h it seems we are not defining FLOAT_SUPPORT.

**Comment by ghaerr** [comment]
> looking at check.h it seems we are not defining FLOAT_SUPPORT.

Yes. I was thinking of turning it on, but we're almost out of space, and, for the moment, didn't want to introduce even more problems than we already have. There is an option to recognize, floating point keywords like 'float' and give an error, but that option seems to have been removed.

At this point, I don't suggest we try getting involved into C86 floating point support. I will help with whatever problem you find in nanoprintf.h if you can find it though, as that could be something else. If float or double were seen in nanoprintf.h, C86 should give a compiler error like you showed above, so that is not the problem, it seems.

**Comment by toncho11** [comment]
OK, you are bypassing the libc to directly call ELKS system calls. But then why an OWC compiled ELKS libc object file can not be used in the toolchain? The linker should pick up the functions used in our C program from this object file and generate the final executable. So what are the technical hurdles for this? Or maybe I oversimplify?

**Comment by ghaerr** [comment]
> So it must be a question of compatibility with the toolchain.

Huge compatibility issues, which is why the process "chains" together separate processing programs. There are tons of issues, but the biggest reasons that the OWC libc.a can't be used for C86 are 1) the .a format is incompatible, 2) the object format with the archive is incompatible, and 3) the compiler itself uses a different function calling convention (register vs stack) for calling routines. So it could never work.

We have the same problem with the ia16-elf-gcc toolchain vs OWC, which is why it took 3 months to get that finalized: the .a format, .o/.obj format and calling sequences are all incompatible. With C86, we've introduced a third set of modules, using the dev86-format .o/.a file format, and mixed with the C86 calling sequence. Luckily, NASM was able to produce an AS86-compatible output format so we could use that with LD86.

Since everything is incompatible, we have to first get a toolchain working, and then, unfortunately, recompile the entire C library from top to bottom. The system call source code is also incompatible, but that was put together a few weeks ago in NASM source format. I pulled that from the NCC project which also happened to use NASM ASM format. (Yes, ASM formats are also incompatible between assemblers).

**Comment by toncho11** [comment]
So you want to compile a new libc using the new compiler toolchain? 
Which one would you use as a base? The OWC or gcc-ia16 version?

**Comment by ghaerr** [comment]
> So you want to compile a new libc using the new compiler toolchain?

Well, actually we want to compile the same C library, as there is only one: the ELKS C library. Both the gcc-ia16 and OWC libraries use the same C library source code (almost).

> Which one would you use as a base? The OWC or gcc-ia16 version?

The C library code is almost completely shared between the OWC and IA16 C libraries, but all the ASM language has to be rewritten. The C86 compiler will also use almost all the same C library code as well, and will also require any ASM language routines to be rewritten for a third time for C86/NASM.

The big advantage in using the same C library is that then programs written using one of the three compilers will generally act exactly the same when calling the same routine. In order to pull off this trickery, what has been done is to "hide" the specific compiler details in a few special files, rather than littered as ifdefs all over the C source. In this way, the C library looks the same, but there is some tricky macros (\#defines) which cause allow different compilers to compile the "same" code.

Most of this special code is in <sys/cdefs.h>, which ends up being in elks/include/arch/cdefs.h. Here's what it looks like, note the special handling for GCC/IA16 vs WATCOM. This is where the special handling for C86 will also be added (I've already started but not actually tried building the C library yet with C86):
```
#ifdef __GNUC__
#define noreturn        __attribute__((__noreturn__)) /* don't require <stdnoreturn.h> */
#define stdcall         __attribute__((__stdcall__))
#define restrict        __restrict
#define printfesque(n)  __attribute__((__format__(__gnu_printf__, n, n + 1)))
#define noinstrument    __attribute__((no_instrument_function))
#define CONSTRUCTOR(fn,pri) void fn(void) __attribute__((constructor(pri)))
#define DESTRUCTOR(fn,pri)  void fn(void) __attribute__((destructor(pri)))
#define __wcfar
#define __wcnear
#endif

#ifdef __WATCOMC__
#define noreturn        __declspec(aborts)
#define stdcall         __stdcall
#define restrict        __restrict
#define printfesque(n)
#define noinstrument
#define CONSTRUCTOR(fn,pri) void fn(void);                                  \
                            static struct _rt_init __based(__segname("XI")) \
                                __CONCAT(_ctor,fn) = { fn, pri, 0}
#define DESTRUCTOR(fn,pri)  void fn(void);                                  \
                            static struct _rt_init __based(__segname("YI")) \
                                __CONCAT(_dtor,fn) = { fn, pri, 0}
#define __attribute__(n)
#define __wcfar         __far
#define __wcnear        __near
#endif

#ifdef __C86__
#define __HAS_NO_FLOATS__
#define __far
#define noreturn
#define stdcall
#define restrict
#define printfesque(n)
#define noinstrument
#define CONSTRUCTOR(fn,pri) void fn(void)
#define DESTRUCTOR(fn,pri)  void fn(void)
#define __attribute__(n)
#define __wcfar
#define __wcnear
#endif
```

**Comment by rafael2k** [comment]
Thanks. Indeed, we need this for C86. For now I left in examples/ two examples that compile fine without headers.

**Comment by toncho11** [comment]
I think it will be useful to explain in the readme:

CPP=cpp86
CC=c86
AS=nasm
LD=ld86

The toolchain first uses CPP (a C pre-processor), next is the c86 compiler that generates NASM compatible assembler, then the nasm compiles both your source code and some other required .asm files and in the end the linker links all object files.

**Comment by ghaerr** [comment]
> then the nasm compiles both your source code and some other required .asm files 

Also @rafael2k, we need to rename nasm to nasm86 (and ndisasm -> ndisasm86) for both host-bin and elks-bin, before it gets too confusing. All the ELKS (and host) development tools should end with -86 in their names, making it very easy to identify any tool part of the cross- or native- development toolchain. (The only exception to this is ELKS make, the host version is already make86). With nasm/ndisasm (and make), there exist host- programs with the same name on most systems, and if the path is set incorrectly, the wrong assembler would get used, causing confusion or errors.

I can do this, unless you'd like to :) I haven't pulled your repo changes for the last week or so because I'm knee-deep in the arena-based malloc allocator which should solve our near heap issues, and also have a set of C86 bug fixes that are based on older commits. I hope get PRs for both very soon!

I will make the nasm->nasm86 changes to `ecc` afterward the compiler bug fixes, as some new options may need to take effect.

**Comment by rafael2k** [comment]
Hi @ghaerr ! Please go ahead, I agree with the 86 suffix, it helps any confusion of other toolchain tools.

**Comment by rafael2k** [comment]
ps: I changed my idea with regards to the 86 suffix for native software.

@toncho11, do you already have a working guide or wiki page to link from the 8086-toolchain README?

**Comment by toncho11** [comment]
This is done by @ghaerr here: https://github.com/ghaerr/elks/wiki/Setting-up-the-8086-toolchain-(C86-compiler-and-tools)
But it is missing explanation on: [buildc86.sh](https://github.com/ghaerr/elks/blob/master/buildc86.sh) and [copyc86.sh](https://github.com/ghaerr/elks/blob/master/copyc86.sh)

Also I think we should relelase both a zip with the current version and bootable elks images. It is a lot of work to set up the the toolbox and people might not have the patience, so better provide a images until we integrate it as part of ELKS.

**Comment by rafael2k** [comment]
I agree with this target - zip/tar.gz with all the needed parts of the toolchain to compile software, plus a bootable (HD I think, as floppy might not be enough) ELKS image with the toolchain installed and ready to use (in correct paths).

**Comment by rafael2k** [comment]
> > zip/tar.gz with all the needed parts of the toolchain to compile software, plus a bootable (HD I think
> > ELKS image with the toolchain installed and ready to use (in correct paths).
> 
> Let me take a pass at this in two phases: I think I can quickly write a script that copies the libc and elks include header and lib.a files from ELKS along with the toolchain binaries from 8086-toolchain onto a 32MB MINIX image build. The second step would be to produce a .tar file to be extracted within an operating ELKS environment which does the same thing. The latter would have to be an uncompress tar file since that's all that ELKS can internally handle.

Yes, I think that would be perfect: 32MB Minix HD image with the toolchain, and later on, a tar with the toolchain ready to untarred (btw, is there any issue in having gzip in ELKS, or just nobody ported yet?).

**Comment by ghaerr** [comment]
> is there any issue in having gzip in ELKS, or just nobody ported yet?)

Gzip itself likely won't port - there are extremely few compressors/decompressors that will run in 16-bits/64K.

However we do have elkscmd/misc_utils/compress.c which is compatible with Linux and macOS compress. This is used by ELKS `man` to decompress man pages on the fly (using fork/exec `compress -d`). We could potentially use 
`compress -d devkit.tar.z | tar xvf -` or something like that to decompress larger images. This would be worth it if the size of the devkit is larger than a floppy uncompressed.
